core/src/main/java/org/bitcoinj/wallet/AnyDeterministicKeyChain.java

```diff
@@ -0,0 +1,1635 @@
+ * sufficient information about the account key to create a watching chain via
+ * {@link KeyFactory#deserializeB58(IDeterministicKey, String, NetworkParameters)}
+ * (with null as the first parameter) and then
+ * {@link Builder#watch(IDeterministicKey)}.</p>
+ *
+ * <p>This class builds on {@link AnyDeterministicHierarchy} and
+ * {@link IDeterministicKey} by adding support for serialization to and from protobufs,
+ * and encryption of parts of the key tree. Internally it arranges itself as per the BIP 32 spec, with the seed being
+ * used to derive a master key, which is then used to derive an account key, the account key is used to derive two
+ * child keys called the <i>internal</i> and <i>external</i> parent keys (for change and handing out addresses respectively)
+ * and finally the actual leaf keys that users use hanging off the end. The leaf keys are special in that they don't
+ * internally store the private part at all, instead choosing to rederive the private key from the parent when
+ * needed for signing. This simplifies the design for encrypted key chains.</p>
+ *
+ * <p>The key chain manages a <i>lookahead zone</i>. This zone is required because when scanning the chain, you don't
+ * know exactly which keys might receive payments. The user may have handed out several addresses and received payments
+ * on them, but for latency reasons the block chain is requested from remote peers in bulk, meaning you must
+ * "look ahead" when calculating keys to put in the Bloom filter. The default lookahead zone is 100 keys, meaning if
+ * the user hands out more than 100 addresses and receives payment on them before the chain is next scanned, some
+ * transactions might be missed. 100 is a reasonable choice for consumer wallets running on CPU constrained devices.
+ * For industrial wallets that are receiving keys all the time, a higher value is more appropriate. Ideally DKC and the
+ * wallet would know how to adjust this value automatically, but that's not implemented at the moment.</p>
+ *
+ * <p>In fact the real size of the lookahead zone is larger than requested, by default, it's one third larger. This
+ * is because the act of deriving new keys means recalculating the Bloom filters and this is an expensive operation.
+ * Thus, to ensure we don't have to recalculate on every single new key/address requested or seen we add more buffer
+ * space and only extend the lookahead zone when that buffer is exhausted. For example with a lookahead zone of 100
+ * keys, you can request 33 keys before more keys will be calculated and the Bloom filter rebuilt and rebroadcast.
+ * But even when you are requesting the 33rd key, you will still be looking 100 keys ahead.
+ * </p>
+ *
+ * @author Andreas Schildbach
+ */
+@SuppressWarnings("PublicStaticCollectionField")
+public class AnyDeterministicKeyChain implements IEncryptableKeyChain {
+    private static final Logger log = LoggerFactory.getLogger(AnyDeterministicKeyChain.class);
```

Comment:
Another copied class.  For now I want to keep `DeterministicKeyChain` untouched since it is used for payments.